src/logics/ArithLogic.cc

```diff
@@ -341,9 +342,15 @@ PTRef polyToPTRef(ArithLogic & logic, poly_t const & poly) {
 }
 
 Logic::SubstMap collectSingleEqualitySubstitutions(ArithLogic & logic, std::vector<poly_t> & zeroPolynomials) {
-    Logic::SubstMap substitutions;
-    std::unordered_map<PTRef, std::vector<std::size_t>, PTRefHash> varToPolyIndices;
+    // MB: We enforce order to ensure that later-created terms are processed first.
+    //     This ensures that from an equality "f(x) = x" we get a substitution "f(x) -> x" and not the other way
+    //     around, which would cause infinite cycle in transitive closure
+    struct PTRefGreaterThan {
+        bool operator()(PTRef first, PTRef second) const { return first.x > second.x; }
+    };
```

Comment:
Isn't there already a greater than operator for `PTRef`?  But in any case the locality of the definition here might actually be a benefit.  I'm fine with this.

Reply:
I am not aware of it.

As the next step, I actually want to clean up the comparison operators for `PTRef`, because in the class, there is `operator<`, which however, compares the ids with `>`, so it is quite confusing.